File: node/src/scanners/regex-scanner.ts
import { spawnSync } from "child_process";
import fs from "fs";
import path from "path";
import { PatternEngine, PatternMatch, Pattern } from "../core/pattern-engine.js";
import { DEFAULT_SECRET_PATTERNS } from "./secret-patterns.js";
import { loadCustomPatterns } from "../core/custom-patterns.js";

export interface ScanResult {
  file: string;
  matches: PatternMatch[];
}

export class RegexScanner {
  private engine: PatternEngine;

  constructor(customPatterns?: Array<{ name: string; regex: string; severity: string }>) {
    const patterns: Pattern[] = [...DEFAULT_SECRET_PATTERNS, ...loadCustomPatterns()];
    if (customPatterns) {
      for (const cp of customPatterns) {
        patterns.push({
          name: cp.name,
          regex: cp.regex,
          severity: cp.severity as Pattern["severity"],
        });
      }
    }
    this.engine = new PatternEngine(patterns);
  }

  /**
   * Scan a single file for secrets. Suppression is applied at the scan
   * command boundary (engine-agnostic), not here.
   */
  scanFile(filePath: string): ScanResult {
    try {
      const content = fs.readFileSync(filePath, "utf-8");
      const matches = this.engine.scanWithPosition(content);
      return { file: filePath, matches };
    } catch (e) {
      return { file: filePath, matches: [] };
    }
  }

  /**
   * Scan multiple files
   */
  scanFiles(filePaths: string[]): ScanResult[] {
    const results: ScanResult[] = [];

    for (const filePath of filePaths) {
      const result = this.scanFile(filePath);
      if (result.matches.length > 0) {
        results.push(result);
      }
    }

    return results;
  }

  /**
   * Scan a directory recursively
   */
  scanDirectory(dirPath: string, options?: {
    exclude?: string[];
    excludePaths?: string[];
    maxDepth?: number;
    /**
     * Honor `.gitignore` (and `.git/info/exclude`, global excludes file,
     * nested .gitignores) when filtering scanned files. Default true.
     * Implemented by shelling out to `git check-ignore --stdin --no-index`
     * inside the scan root's git work tree — exact git semantics, zero deps.
     * Silently falls back to no-op when the scan root is outside a git repo
     * or git is missing.
     */
    respectGitignore?: boolean;
  }): ScanResult[] {
    const exclude = options?.exclude || [
      "node_modules",
      ".git",
      "dist",
      "build",
      ".next",
      "coverage",
      ".vscode",
      ".idea",
      // Vendored / virtual-env / generated dirs that cause false positives
      "vendor",
      ".venv",
      "venv",
      "__pycache__",
      ".tox",
      ".mypy_cache",
      ".pytest_cache",
      "results",
      ".terraform",
      "bower_components"
    ];

    // Merge policy excludePaths into the exclude list
    if (options?.excludePaths) {
      for (const ep of options.excludePaths) {
        // Strip trailing slashes for directory name matching
        const cleaned = ep.replace(/\/+$/, "");
        if (!exclude.includes(cleaned)) {
          exclude.push(cleaned);
        }
      }
    }

    const files = this.walkDirectory(dirPath, exclude, options?.maxDepth || 10);
    const filtered = options?.respectGitignore === false
      ? files
      : filterGitIgnored(files, dirPath);
    return this.scanFiles(filtered);
  }

  /**
   * Scan text content directly
   */
  scanText(text: string): PatternMatch[] {
    return this.engine.scan(text);
  }

  /**
   * Scan a single line at a known file line number (git diff + side).
   */
  scanLine(text: string, lineNumber: number): PatternMatch[] {
    return this.engine.scanWithPosition(text).map((m) => ({
      ...m,
      line: lineNumber,
    }));
  }

  /**
   * Redact secrets from text
   */
  redact(text: string): string {
    return this.engine.redactText(text);
  }

  /**
   * Check if text contains secrets
   */
  hasSecrets(text: string): boolean {
    return this.engine.hasMatches(text);
  }

  /**
   * Walk directory and collect file paths
   */
  private walkDirectory(dir: string, exclude: string[], maxDepth: number, currentDepth = 0): string[] {
    if (currentDepth >= maxDepth) {
      return [];
    }

    const files: string[] = [];

    try {
      const entries = fs.readdirSync(dir, { withFileTypes: true });

      for (const entry of entries) {
        // Skip symlinks to prevent traversal outside intended scope
        if (entry.isSymbolicLink()) {
          continue;
        }

        const fullPath = path.join(dir, entry.name);

        // Skip excluded directories
        if (exclude.includes(entry.name)) {
          continue;
        }

        if (entry.isDirectory()) {
          files.push(...this.walkDirectory(fullPath, exclude, maxDepth, currentDepth + 1));
        } else if (entry.isFile()) {
          // Skip binary files by extension
          if (!this.isBinaryFile(entry.name)) {
            files.push(fullPath);
          }
        }
      }
    } catch (e) {
      // Skip directories we can't read
    }

    return files;
  }

  /**
   * Check if file is likely binary based on extension
   */
  private isBinaryFile(filename: string): boolean {
    const binaryExtensions = [
      ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".ico",
      ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
      ".zip", ".tar", ".gz", ".rar", ".7z",
      ".exe", ".dll", ".so", ".dylib",
      ".mp3", ".mp4", ".avi", ".mov",
      ".woff", ".woff2", ".ttf", ".eot",
      ".pyc", ".class", ".o", ".a"
    ];

    const ext = path.extname(filename).toLowerCase();
    return binaryExtensions.includes(ext);
  }
}

/**
 * Filter out files that git would ignore relative to `scanRoot`.
 *
 * Uses `git check-ignore --stdin --no-index` inside the scan root's work
 * tree so every gitignore semantic git itself supports (nested .gitignores,
 * negations, .git/info/exclude, the configured global excludes file, the
 * `gitignore` pattern grammar) is honored exactly. Zero new dependencies —
 * if git isn't installed or the scan root sits outside a work tree, returns
 * the input unchanged.
 *
 * Batched: all candidate paths are piped through one subprocess.
 */
function filterGitIgnored(files: string[], scanRoot: string): string[] {
  if (files.length === 0) return files;

  // Locate the git work tree that owns scanRoot. Bail out (no filter) if
  // scanRoot is outside a repo or git is missing — the user's `rafter
  // secrets <dir>` invocation against a non-repo should not error.
  const probe = spawnSync("git", ["-C", scanRoot, "rev-parse", "--show-toplevel"], {
    encoding: "utf-8",
    stdio: ["ignore", "pipe", "ignore"],
    timeout: 5000,
  });
  if (probe.status !== 0) return files;
  const workTree = probe.stdout.trim();
  if (!workTree) return files;

  // Send the candidate file list to `git check-ignore --stdin`. It echoes
  // back (on stdout) only the paths that match a gitignore rule. Use null
  // delimiters (`-z`) on input and output so paths with spaces, newlines, or
  // any other shell metacharacter survive intact. `--no-index` makes git
  // evaluate rules without consulting the index (so an untracked file in a
  // freshly-cloned repo is still classified).
  //
  // The paths in `files` are absolute. `git check-ignore` accepts paths
  // relative to the work tree OR absolute paths within it; we pass through
  // unchanged.
  const stdin = files.join("\0") + "\0";
  const result = spawnSync(
    "git",
    ["-C", workTree, "check-ignore", "--stdin", "--no-index", "-z"],
    {
      input: stdin,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "ignore"],
      timeout: 30_000,
      maxBuffer: 64 * 1024 * 1024,
    }
  );
  // Exit codes: 0 = at least one path matched; 1 = no paths matched;
  // 128 = error (corrupt repo, bad flag, etc.). Treat 128 as "no filter" so
  // a broken environment doesn't break a scan.
  if (result.status !== 0 && result.status !== 1) return files;

  const ignored = new Set<string>();
  if (result.stdout) {
    for (const p of result.stdout.split("\0")) {
      if (p) ignored.add(p);
    }
  }
  if (ignored.size === 0) return files;
  return files.filter((f) => !ignored.has(f));
}

File: python/rafter_cli/scanners/regex_scanner.py
"""Regex-based secret scanner."""
from __future__ import annotations

import os
import subprocess
from dataclasses import dataclass, field
from pathlib import Path

from ..core.pattern_engine import Pattern, PatternEngine, PatternMatch
from .secret_patterns import DEFAULT_SECRET_PATTERNS
from ..core.custom_patterns import load_custom_patterns


@dataclass
class ScanResult:
    file: str
    matches: list[PatternMatch] = field(default_factory=list)


BINARY_EXTENSIONS = frozenset([
    ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".ico",
    ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
    ".zip", ".tar", ".gz", ".rar", ".7z",
    ".exe", ".dll", ".so", ".dylib",
    ".mp3", ".mp4", ".avi", ".mov",
    ".woff", ".woff2", ".ttf", ".eot",
    ".pyc", ".class", ".o", ".a",
])

DEFAULT_EXCLUDE = [
    "node_modules", ".git", "dist", "build",
    ".next", "coverage", ".vscode", ".idea",
    "__pycache__", ".mypy_cache", ".ruff_cache",
]


class RegexScanner:
    def __init__(self, custom_patterns: list[dict] | None = None):
        patterns: list[Pattern] = list(DEFAULT_SECRET_PATTERNS) + load_custom_patterns()
        if custom_patterns:
            for cp in custom_patterns:
                patterns.append(Pattern(
                    name=cp.get("name", "Custom"),
                    regex=cp.get("regex", ""),
                    severity=cp.get("severity", "high"),
                ))
        self._engine = PatternEngine(patterns)
        self._skipped: list[dict[str, str]] = []

    def scan_file(self, file_path: str) -> ScanResult:
        # Suppression is applied at the scan command boundary (engine-agnostic).
        try:
            content = Path(file_path).read_text(errors="ignore")
        except (OSError, UnicodeDecodeError) as exc:
            # A path we could not read is NOT a path with no secrets in it.
            # Record it so the caller can say so; an empty match list on its
            # own is indistinguishable from a clean file.
            self._skipped.append(
                {"path": file_path, "reason": getattr(exc, "strerror", None) or type(exc).__name__}
            )
            return ScanResult(file=file_path)
        matches = self._engine.scan_with_position(content)
        return ScanResult(file=file_path, matches=matches)

    def scan_files(self, file_paths: list[str]) -> list[ScanResult]:
        results: list[ScanResult] = []
        for fp in file_paths:
            r = self.scan_file(fp)
            if r.matches:
                results.append(r)
        return results

    def scan_directory(
        self,
        dir_path: str,
        exclude_paths: list[str] | None = None,
        max_depth: int = 10,
        respect_gitignore: bool = True,
    ) -> list[ScanResult]:
        exclude = list(DEFAULT_EXCLUDE)
        if exclude_paths:
            for ep in exclude_paths:
                cleaned = ep.rstrip("/")
                if cleaned not in exclude:
                    exclude.append(cleaned)

        files = self._walk(dir_path, exclude, max_depth)
        if respect_gitignore:
            files = _filter_gitignored(files, dir_path)
        return self.scan_files(files)

    def scan_text(self, text: str) -> list[PatternMatch]:
        return self._engine.scan(text)

    def scan_line(self, text: str, line_number: int) -> list[PatternMatch]:
        """Scan a single line at a known file line number (git diff + side)."""
        return [
            PatternMatch(
                pattern=m.pattern,
                match=m.match,
                line=line_number,
                column=m.column,
                redacted=m.redacted,
                engines=m.engines,
            )
            for m in self._engine.scan_with_position(text)
        ]

    def redact(self, text: str) -> str:
        return self._engine.redact_text(text)

    def has_secrets(self, text: str) -> bool:
        return self._engine.has_matches(text)

    # ------------------------------------------------------------------

    def take_skipped(self) -> list[dict[str, str]]:
        """Paths this scanner could not read, drained (and cleared) by the caller.

        Both skip sites — an unreadable file and an unreadable directory — used
        to be silent, so ``rafter secrets <unreadable>`` returned exit 0 with
        empty results, byte-identical to a genuinely clean scan. The directory
        a scanner cannot enter is exactly where an unnoticed credential is most
        likely to be.
        """
        out = self._skipped
        self._skipped = []
        return out

    def _walk(self, directory: str, exclude: list[str], max_depth: int, depth: int = 0) -> list[str]:
        if depth >= max_depth:
            return []
        files: list[str] = []
        try:
            for entry in os.scandir(directory):
                if entry.name in exclude:
                    continue
                # Skip symlinks to prevent traversal outside intended scope
                if entry.is_symlink():
                    continue
                if entry.is_dir(follow_symlinks=False):
                    files.extend(self._walk(entry.path, exclude, max_depth, depth + 1))
                elif entry.is_file(follow_symlinks=False):
                    ext = os.path.splitext(entry.name)[1].lower()
                    if ext not in BINARY_EXTENSIONS:
                        files.append(entry.path)
        except PermissionError as exc:
            # A directory we can't read is not an empty directory — record it
            # so the caller can report coverage honestly.
            self._skipped.append(
                {"path": directory, "reason": getattr(exc, "strerror", None) or "PermissionError"}
            )
        return files


def _filter_gitignored(files: list[str], scan_root: str) -> list[str]:
    """Drop files git would ignore relative to ``scan_root``.

    Shells out to ``git check-ignore --stdin --no-index -z`` inside the scan
    root's work tree so every gitignore semantic git supports (nested
    .gitignores, negations, .git/info/exclude, the configured global excludes
    file, the full pattern grammar) is honored exactly. Zero new
    dependencies. Bails out silently — returning the input unchanged — when
    git is missing or the scan root sits outside a work tree.
    """
    if not files:
        return files
    # Locate the work tree that owns scan_root.
    try:
        probe = subprocess.run(
            ["git", "-C", scan_root, "rev-parse", "--show-toplevel"],
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return files
    if probe.returncode != 0:
        return files
    work_tree = probe.stdout.strip()
    if not work_tree:
        return files

    # Batch every candidate through one `git check-ignore --stdin` call.
    # Null-delimited I/O preserves paths containing newlines / spaces.
    stdin = ("\0".join(files) + "\0").encode("utf-8")
    try:
        result = subprocess.run(
            ["git", "-C", work_tree, "check-ignore", "--stdin", "--no-index", "-z"],
            input=stdin,
            capture_output=True,
            timeout=30,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return files
    # Exit codes: 0 = at least one path ignored; 1 = none; 128 = error.
    # Anything else => abandon the filter rather than break the scan.
    if result.returncode not in (0, 1):
        return files
    ignored = {p.decode("utf-8") for p in result.stdout.split(b"\0") if p}
    if not ignored:
        return files
    return [f for f in files if f not in ignored]
